tikv/write.go

```diff
@@ -214,114 +177,212 @@ func (w *writeLockWorker) run() {
+	writer *dbWriter
 }
 
-func (w *rollbackGCWorker) run() {
-	store := w.store
-	defer store.wg.Done()
+func (w rollbackGCWorker) run() {
+	defer w.writer.wg.Done()
 	ticker := time.Tick(time.Minute)
+	rollbackStore := w.writer.bundle.RollbackStore
 	for {
 		select {
-		case <-store.closeCh:
+		case <-w.writer.closeCh:
 			return
 		case <-ticker:
 		}
-		lockBatch := newWriteLockBatch(new(requestCtx))
-		it := store.rollbackStore.NewIterator()
-		latestTS := store.getLatestTS()
+		lockBatch := &writeLockBatch{}
+		it := rollbackStore.NewIterator()
+		latestTS := w.writer.getLatestTS()
 		for it.SeekToFirst(); it.Valid(); it.Next() {
 			ts := mvcc.DecodeRollbackTS(it.Key())
 			if tsSub(latestTS, ts) > time.Minute {
 				lockBatch.rollbackGC(safeCopy(it.Key()))
 			}
 			if len(lockBatch.entries) >= 1000 {
-				store.writeLocks(lockBatch)
+				lockBatch.wg.Add(1)
+				w.writer.lockCh <- lockBatch
+				lockBatch.wg.Wait()
 				lockBatch.entries = lockBatch.entries[:0]
 			}
 		}
 		if len(lockBatch.entries) == 0 {
 			continue
 		}
-		store.writeLocks(lockBatch)
+		lockBatch.wg.Add(1)
+		w.writer.lockCh <- lockBatch
+		lockBatch.wg.Wait()
 	}
 }
 
-type lockEntryHdr struct {
-	keyLen uint32
-	valLen uint32
+type dbWriter struct {
+	bundle    *mvcc.DBBundle
+	safePoint *SafePoint
+	dbCh      chan<- *writeDBBatch
+	lockCh    chan<- *writeLockBatch
+	wg        sync.WaitGroup
+	closeCh   chan struct{}
+	latestTS  uint64
 }
 
-func (store *MVCCStore) dumpMemLocks() error {
-	tmpFileName := store.dir + "/lock_store.tmp"
-	f, err := os.OpenFile(tmpFileName, os.O_CREATE|os.O_TRUNC|os.O_RDWR, 0666)
-	if err != nil {
-		return errors.Trace(err)
-	}
-	writer := bufio.NewWriter(f)
-	cnt := 0
-	it := store.lockStore.NewIterator()
-	hdrBuf := make([]byte, 8)
-	hdr := (*lockEntryHdr)(unsafe.Pointer(&hdrBuf[0]))
-	for it.SeekToFirst(); it.Valid(); it.Next() {
-		hdr.keyLen = uint32(len(it.Key()))
-		hdr.valLen = uint32(len(it.Value()))
-		writer.Write(hdrBuf)
-		writer.Write(it.Key())
-		writer.Write(it.Value())
-		cnt++
+func NewDBWriter(bundle *mvcc.DBBundle, safePoint *SafePoint) mvcc.DBWriter {
+	return &dbWriter{
+		bundle:    bundle,
+		safePoint: safePoint,
+		closeCh:   make(chan struct{}, 0),
 	}
-	err = writer.Flush()
-	if err != nil {
-		return errors.Trace(err)
+}
+
+func (writer *dbWriter) Open() {
+	writer.wg.Add(3)
+
+	dbCh := make(chan *writeDBBatch, batchChanSize)
+	writer.dbCh = dbCh
+	go writeDBWorker{
+		batchCh: dbCh,
+		writer:  writer,
+	}.run()
+
+	lockCh := make(chan *writeLockBatch, batchChanSize)
+	writer.lockCh = lockCh
+	go writeLockWorker{
+		batchCh: lockCh,
+		writer:  writer,
+	}.run()
+
+	go rollbackGCWorker{
+		writer: writer,
+	}.run()
+}
+
+func (writer *dbWriter) Close() {
+	close(writer.closeCh)
+	writer.wg.Wait()
+}
+
+func (writer *dbWriter) Write(batch mvcc.WriteBatch) error {
+	wb := batch.(*writeBatch)
+	if len(wb.dbBatch.entries) > 0 {
+		// We must delete lock after commit succeed, or there will be inconsistency.
```

Comment:
I think this line comment can be moved to L275?

Reply:
ok